### PR TITLE
Support late player attaching

### PR DESCRIPTION
**Pull request by stonko1994**
### Problem

There are certain scenarios where VST tracking should start before a Player instance is created. Currently there is no way to start a Conviva Session without a player instance.

### Solution

Extending the API to allow session initialization without a Player instance and adding a possibility to attach a Player instance at a later point in the life-cycle of the session.

### Notes

Manual session initialization API already existed and still needs to be used to start VST tracking.

### Manual testing instructions
I extended the Sample with a new `Start session` button. To test this, start the sample but first hit the `Release` button to stop and destroy the player/session, which automatically starts. (Of comment out `setupBitmovinPlayer` in `viewDidLoad`).
- Tap `Start Session` button
- Tap `Create` button
- Double check in touchstone that the duration between those two actions is properly tracked as VST.

### Checklist
- [x] I added an update to `CHANGELOG.md` file
- [x] I ran all the tests in the project and they succeed

### CODE REVIEW

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java [hunk 1]
```diff
@@ -299,6 +336,30 @@ public void reportAppBackgrounded() {
...
+     * without `player` if you plan to attach a `Player` instance later in the life-cycle.
+     */
+    public void attachPlayer(@NonNull Player player) {
+        if (this.playerAdapter != null) {
+            Log.w(TAG, "There is already a Player instance attached! Ignoring new Player instance.");
+            return;
+        }
+
+        if (player.getSource() != null) {
+            Log.w(TAG, "Player already has a source loaded. Please call setPlayer before loading the source.");
```
Comment (strangesource) [inline conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:353]:
```suggestion
            Log.w(TAG, "Player already has a source loaded. Please provide the player instance before loading a source.");
```

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ssai/DefaultSsaiApi.java [hunk 2]
```diff
@@ -71,12 +81,16 @@ public void reportAdStarted(AdInfo adInfo) {
             Log.d(TAG, "No server side ad break active");
             return;
         }
+        if (playerAdapter == null) {
+            Log.d(TAG, "Player not yet set. Cannot report ad started.");
```
Comment (strangesource) [inline conviva/src/main/java/com/bitmovin/analytics/conviva/ssai/DefaultSsaiApi.java:85]: Should probably be a warning.

File: conviva/src/test/kotlin/com/bitmovin/analytics/conviva/LatePlayerAttachingTest.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    companion object {
+        @JvmStatic
+        @BeforeClass
+        fun beforeClass() {
+            mockkStatic(Log::class)
+            every { Log.v(any(), any()) } returns 0
+            every { Log.d(any(), any()) } returns 0
+            every { Log.i(any(), any()) } returns 0
+            every { Log.e(any(), any()) } returns 0
+            every { Log.w(any(), any<String>()) } returns 0
```
Comment (strangesource) [inline conviva/src/test/kotlin/com/bitmovin/analytics/conviva/LatePlayerAttachingTest.kt:85]: NIT I think this is the third time we are mocking the logs, maybe worth to extract into a helper function.

**Review by strangesource — COMMENTED** [review]
Looking good, just some NIT from my side.

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java (hunk 1 above)
Comment (stonko1994) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java]: [d26112a](https://github.com/bitmovin/bitmovin-player-android-analytics-conviva/pull/80/commits/d26112a7be48b6ae2d8040cacb3b38594c1a3467)

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ssai/DefaultSsaiApi.java (hunk 2 above)
Comment (stonko1994) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ssai/DefaultSsaiApi.java]: [e709ddd](https://github.com/bitmovin/bitmovin-player-android-analytics-conviva/pull/80/commits/e709dddb6e463aa615cd3e14d040205b060530e2)

File: conviva/src/test/kotlin/com/bitmovin/analytics/conviva/LatePlayerAttachingTest.kt (hunk 3 above)
Comment (stonko1994) [reply conviva/src/test/kotlin/com/bitmovin/analytics/conviva/LatePlayerAttachingTest.kt:85]: Do you have a hint on how to do that in an idiomatic way? 🤔 Is it enough to create a helper class which is called form those companion objects or is there a better approach to include this into tests 🤔

Comment (strangesource) [reply conviva/src/test/kotlin/com/bitmovin/analytics/conviva/LatePlayerAttachingTest.kt:85]: I'd suggest a top level function similar to 
```
      fun mockLogging() {
            mockkStatic(Log::class)
            every { Log.v(any(), any()) } returns 0
            every { Log.d(any(), any()) } returns 0
            every { Log.i(any(), any()) } returns 0
            every { Log.e(any(), any()) } returns 0
            every { Log.w(any(), any<String>()) } returns 0
        }
```

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ssai/WithEventEmitter.java
Comment (strangesource) [inline conviva/src/main/java/com/bitmovin/analytics/conviva/ssai/WithEventEmitter.java:1]: Should probably not be in the ssai package

Comment (stonko1994) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ssai/WithEventEmitter.java:1]: [c6d4ab3](https://github.com/bitmovin/bitmovin-player-android-analytics-conviva/pull/80/commits/c6d4ab366a003e77c98067cf109fa3a8f3ff7ca1)

File: conviva/src/test/kotlin/com/bitmovin/analytics/conviva/LatePlayerAttachingTest.kt (hunk 3 above)
Comment (stonko1994) [reply conviva/src/test/kotlin/com/bitmovin/analytics/conviva/LatePlayerAttachingTest.kt]: [37ef1d6](https://github.com/bitmovin/bitmovin-player-android-analytics-conviva/pull/80/commits/37ef1d65db7f4a7e7b4ecde3f4526b09cd00707c)

**Review by strangesource — APPROVED** [review]

**Pull request merged**